keps/0035-json-schema-export.md

```diff
@@ -0,0 +1,265 @@
+
+### JSON-Schema export
+
+Extend the `kudo package list parameters` with an option `--output json-schema` to describe the parameters as JSON-Schema.
+
+Alternatively, add a new command `kudo package list json-schema`, which then could be used with `--output json` or `--output yaml` to export the JSON-Schema in both variants.
+
+The output must include the additional attributes as well as an attribute `listName` that includes the `name` of the parameter to allow a user to generate the correct list of parameters. The group should describe a top-level group. 
+Example:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+groups:
+  - name: "Node"
+    title: "Node Configuration"
+    description: "This section contains parameters that are related to the Node"
+parameters:
+  - name: NODE_COUNT
+    title: "Node Count"
+    description: "Number of Cassandra nodes."
+    type: number
+    default: "3"
+  - name: NODE_CPU_MC
+    title: "Requested CPU"
+    hint: "CPU request (in millicores) for the Cassandra node containers."
+    default: "1000"
+    type: number
+    group: "Node"
+    required: true
+  - name: NODE_MEM_MIB
+    title: "Requested Memory"
+    hint: "Memory request (in MiB) for the Cassandra node containers."
+    type: number
+    default: "4096"
+    group: "Node"
+    required: true
+  - name: NODE_DOCKER_IMAGE
+    title: "Docker Image"
+    hint: "Cassandra node Docker image"
+    description: "The docker image to be used for the Node. Must be one of the mesosphere/cassandra images, as it contains customizations."
+    default: "mesosphere/cassandra:3.11.6-1.0.1"
+    group: "Node"
+    advanced: true
+  - name: NODE_DOCKER_IMAGE_PULL_POLICY
+    title: "Docker Image Pull Policy"
+    hint: "Cassandra node Docker image pull policy"
+    description: "The Docker image pull policy. Can be changed, for example if the image is pre-loaded on the Kubernetes nodes."
+    default: "Always"
+    type: "string"
+    enum: 
+      - "Always"
+      - "Never"
+      - "IfNotPresent"
+    group: "Node"
+    advanced: true
+```
+
+The exported JSON-Schema for this parameter definition should look as following:
+
+```json
+{
+  "title": "cassandra",
+  "type": "object",
+  "properties": {
+    "NODE_COUNT": {
+      "type": "number",
+      "title": "Node Count",
+      "description": "Number of Cassandra nodes.",
+      "default": "3"
+    },
+    "Node": {
+      "type": "object",
+      "title": "Node Configuration",
+      "description": "This section contains parameters that are related to the Node",
+      "required": [
+        "NODE_CPU_MC",
+        "NODE_MEM_MIB"
+      ],
+      "properties":  {
+        "NODE_CPU_MC": {
```

Comment:
When submitting a UI form generated from this json-schema, this parameter value will be nested and look smth. like:
```
Node: {
  NODE_CPU_MC: 2000
}
```
Will kudoctl now understand nested parameters e.g. `Node.NODE_CPU_MC` or will we expect UI only use the parameter `names`?

Reply:
No, KUDO will not understand nested parameters. The UI is expected to generate a list of values from the `listName` attribute in each parameter. Using the `name` would be an alternative....

Reply:
I've added a small paragraph about that